src/coprocessor/statistics/analyze.rs

```diff
@@ -391,41 +391,36 @@ impl<S: Snapshot, F: KvFormat> RowSampleBuilder<S, F> {
                 is_drained = result.is_drained?;
 
                 let columns_slice = result.physical_columns.as_slice();
-
+                let mut column_vals: Vec<Vec<u8>> = vec![vec![]; self.columns_info.len()];
```

Comment:
I believe you can use `Vec::with_capacity` to achieve the same goal without doing a lot of index access in this PR? 🤪

Reply:
Do you mean like the following code?
```
                for logical_row in &result.logical_rows {
                    let mut column_vals: Vec<Vec<u8>> = Vec::with_capacity(self.columns_info.len());
                    let mut collation_key_vals: Vec<Vec<u8>> = Vec::with_capacity(self.columns_info.len());
                    for i in 0..self.columns_info.len() {
                        let mut val = Vec::with_capacity(get_bytes_len_after_encoding());
```
`Vec::with_capacity` can reduce mem allocation times. However, when we handle each `logical_row`, we still need to allocate mem for `column_vals` and `collation_key_vals`.

Basically, the PR only needs to allocate mem for `column_vals` and `collation_key_vals` once. It doesn't need to allocate any mem in `for logical_row in &result.logical_rows` loop if all the columns have fixed sizes. Mem allocation only happens when some column is string-type and its value of the current row is longer than that of all previous rows.

Reply:
The following micro bench shows the difference:
```
    fn prepare_encode_column_arguments() -> (Vec<tipb::ColumnInfo>, Vec<LazyBatchColumn>) {
        let mut columns_info = Vec::new();
        for i in 1..4 {
            let mut col_info = tipb::ColumnInfo::default();
            col_info.set_column_id(i as i64);
            col_info.as_mut_accessor().set_tp(FieldTypeTp::LongLong);
            columns_info.push(col_info);
        }
        let mut columns_slice = Vec::new();
        for i in 0..3 {
            let mut col = LazyBatchColumn::decoded_with_capacity_and_tp(1, EvalType::Int);
            col.mut_decoded().push_int(Some(i));
            columns_slice.push(col)
        }
        (columns_info, columns_slice)
    }

    fn encode_column_without_capacity(columns_info: &Vec<tipb::ColumnInfo>, columns_slice: &Vec<LazyBatchColumn>) {
        let mut column_vals = Vec::new();
        for i in 0..columns_info.len() {
            let mut val = vec![];
            columns_slice[i]
                .encode(0, &columns_info[i], &mut EvalContext::default(), &mut val)
                .unwrap();
            column_vals.push(val);
        }
    }

    fn encode_column_with_capacity(columns_info: &Vec<tipb::ColumnInfo>, columns_slice: &Vec<LazyBatchColumn>) {
        let mut column_vals = Vec::with_capacity(columns_info.len());
        for i in 0..columns_info.len() {
            let mut val = Vec::with_capacity(9); // The length of output of encoding one int value is 9.
            columns_slice[i]
                .encode(0, &columns_info[i], &mut EvalContext::default(), &mut val)
                .unwrap();
            column_vals.push(val);
        }
    }

    fn encode_column_without_allocation(columns_info: &Vec<tipb::ColumnInfo>, columns_slice: &Vec<LazyBatchColumn>, column_vals: &mut Vec<Vec<u8>>) {
        for i in 0..columns_info.len() {
            column_vals[i].clear();
            columns_slice[i]
                .encode(0, &columns_info[i], &mut EvalContext::default(), &mut column_vals[i])
                .unwrap();
        }
    }

    #[bench]
    fn bench_encode_column_without_capacity(b: &mut test::Bencher) {
        let (columns_info, columns_slice) = prepare_encode_column_arguments();
        b.iter(|| {
            encode_column_without_capacity(&columns_info, &columns_slice);
        });
    }

    #[bench]
    fn bench_encode_column_with_capacity(b: &mut test::Bencher) {
        let (columns_info, columns_slice) = prepare_encode_column_arguments();
        b.iter(|| {
            encode_column_with_capacity(&columns_info, &columns_slice);
        });
    }

    #[bench]
    fn bench_encode_column_without_allocation(b: &mut test::Bencher) {
        let (columns_info, columns_slice) = prepare_encode_column_arguments();
        // Pre allocate mem for writing the encoding output. 9 is the length of output of encoding one int value.
        let mut column_vals = vec![vec![0 as u8;9],vec![0 as u8;9], vec![0 as u8;9]];
        b.iter(|| {
            encode_column_without_allocation(&columns_info, &columns_slice, &mut column_vals);
        });
    }
```
```
running 3 tests
test coprocessor::statistics::analyze::benches::bench_encode_column_with_capacity           ... bench:         784 ns/iter (+/- 64)
test coprocessor::statistics::analyze::benches::bench_encode_column_without_allocation      ... bench:         406 ns/iter (+/- 32)
test coprocessor::statistics::analyze::benches::bench_encode_column_without_capacity        ... bench:       1,016 ns/iter (+/- 69)
```

Reply:
![Screen Shot 2023-02-26 at 4 06 08 PM](https://user-images.githubusercontent.com/30385241/221399754-8097ae9f-86d5-492a-a01e-1b55c1592ddf.png)
We can see that in the tikv profile when executing analyze, vec mem allocation takes roughly 20%. Hence, I choose to reuse `column_vals` and `collation_key_vals` to improve performance, though it sacrifices some readability.